driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -1170,55 +1154,170 @@ int getNumPinnedToTransaction() {
         }
     }
 
+    /**
+     * This class maintains threads needed to perform {@link #getAsync(SingleResultCallback)}.
+     */
     @ThreadSafe
-    private static class Workers implements AutoCloseable {
-        private volatile ExecutorService getter;
-        private volatile ExecutorService opener;
+    private static class AsyncWorkManager implements AutoCloseable {
+        private volatile State state;
+        private volatile BlockingQueue<Task> tasks;
         private final Lock lock;
+        @Nullable
+        private ExecutorService worker;
 
-        Workers() {
+        AsyncWorkManager() {
+            state = State.NEW;
+            tasks = new LinkedBlockingQueue<>();
             lock = new StampedLock().asWriteLock();
         }
 
-        Executor getter() {
-            if (getter == null) {
-                lock.lock();
-                try {
-                    if (getter == null) {
-                        getter = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncGetter"));
-                    }
-                } finally {
-                    lock.unlock();
+        void enqueue(final Task task) {
+            lock.lock();
+            try {
+                if (initUnlessClosed()) {
+                    tasks.add(task);
+                    return;
                 }
+            } finally {
+                lock.unlock();
             }
-            return getter;
+            task.failAsClosed();
         }
 
-        Executor opener() {
-            if (opener == null) {
-                lock.lock();
-                try {
-                    if (opener == null) {
-                        opener = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncOpener"));
-                    }
-                } finally {
-                    lock.unlock();
-                }
+        /**
+         * @return {@code false} iff the {@link #state} is {@link State#CLOSED}.
+         */
+        private boolean initUnlessClosed() {
+            boolean result = true;
+            if (state == State.NEW) {
+                worker = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncGetter"));
+                assertNotNull(worker).submit(() -> runAndLogUncaught(this::workerRun));
```

Comment:
Nit: I think this assertion is unnecessary.  Can't we assume that the JDK is implemented to its specification?

Reply:
Hilarious! 😆 I believe I had `worker` marked as `volatile` at some point. Given that it is also `@Nullable`, the assertion was needed to make IDEA happy. Now that the field is plain, IDEA is smart enough to see that it cannot be `null` at this point.

Done.